python-package/lightgbm/dask.py

```diff
@@ -110,12 +110,128 @@ def _train_part(
     else:
         init_score = None
 
+    # construct local eval_set data.
+    local_eval_set = None
+    local_eval_names = None
+    local_eval_sample_weight = None
+    local_eval_class_weight = None
+    local_eval_init_score = None
+    local_eval_group = None
+    n_evals = max([len(x.get('eval_set', [])) for x in list_of_parts])
+    if n_evals:
+
+        has_eval_names = any([x.get('eval_names') is not None for x in list_of_parts])
+        has_eval_sample_weight = any([x.get('eval_sample_weight') is not None for x in list_of_parts])
+        has_eval_class_weight = any([x.get('eval_class_weight') is not None for x in list_of_parts])
+        has_eval_init_score = any([x.get('eval_init_score') is not None for x in list_of_parts])
+
+        local_eval_set = []
+        if has_eval_names:
+            local_eval_names = []
+        if has_eval_sample_weight:
+            local_eval_sample_weight = []
+        if has_eval_class_weight:
+            local_eval_class_weight = []
+        if has_eval_init_score:
+            local_eval_init_score = []
+        if is_ranker:
+            local_eval_group = []
+
+        # consolidate parts of each individual eval component.
+        for i in range(n_evals):
+            x_e = []
+            y_e = []
+            w_e = []
+            g_e = []
+            init_score_e = []
+            for part in list_of_parts:
+
+                if not part.get('eval_set'):
+                    continue
+
+                # possible that not each part contains parts of each individual (X, y) eval set.
+                if i >= len(part['eval_set']):
+                    continue
```

Comment:
So if I understand this, it's saying "if 4 eval sets were given, `i = 4`, and this particular `part["eval_set"]` only has 3 things in it, then move on to the next part"?

Is it possible for this logic to result in mixing of eval sets? Like is this a scenario that could happen:

1. I have 4 eval sets, A, B, C, and D
2. one `part` has `{"eval_sets": [p_A, p_B, p_C]}` and another has `{"eval_sets": [p_A, p_B, p_D]}`
3. as a result, when `i = 2` in this loop, I end up concatenating a piece of C (`p_C`) and a piece of D (`p_D`)

If that's possible, then I think it would be safer to pad the `eval_*` stuff in `parts` with `None`'s so that you can always rely on the positional ordering. That would only work if `_concat()` can handle `None`, something I haven't checked yet.

What do you think?

Reply:
Omg. Padding, this is a great idea! Yeah this is right. If you order your eval_sets from smallest to largest this could happen. I don't think it would happen if they are ordered largest to smallest, in which case each part's first eval_set would definitely come from (X1, y1), and the second from (X2, y2) kind of thing. But yeah, I think padding is a great solution, definitely better than re-ordering the eval_sets before `_train` and then re-sorting validation data artifacts to match the original `eval_set` input ordering.

I think the changes required would be closer to [line 476](https://github.com/microsoft/LightGBM/pull/4101/files#diff-4583a656084ea62b391a061e1c1f533e3e99e1d1a6021a5408032d80c6bdd394R476), I could make sure that each individual `eval_sets[parts_idx]` has the same length, that being `len(eval_sets)` through None padding.

Let me get back to you after I've had a moment to think through what None padding would look like. `_concat` might come out unscathed.

Reply:
Hey I think I've gotten the padding to work, without touching `_concat`. Update to PR coming tomorrow evening once I've cleaned up the code.

Reply:
hey awesome, thanks!

By the way, @jmoralez recently added #4159 which makes the unit tests a LOT faster. We just merged that this morning. So if you merge `master`, you should be able to iterate faster.

Reply:
Omigod @jmoralez very slick!!! Thank you!

Reply:
Padding issue resolved in c1f6e0018.

Each individual `part`'s `eval_set` gets padded with `None`(s) whenever there are no more parts of an individual eval_set component to distribute. So let's say that there are two eval_sets - the first one has 10 partitions and the other has 20. Then parts 11-20 for the second eval_set component will be batched up with 10 None's for padding from the first set. I can add a Lucidchart diagram in the AM.

This way, each part's `eval_set` has the same length, which should make each worker's job of consolidating eval_sets across parts immune from mixing up components' parts.